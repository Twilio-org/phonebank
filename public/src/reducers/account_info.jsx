const userAccountInfo = {
  first_name: null,
  last_name: null,
  email: null,
  phone_number: null,
  is_admin: null
};

export const SET_USER_ACCOUNT_INFO = 'SET_USER_ACCOUNT_INFO';

export function accountInfoReducer(state = userAccountInfo, action) {
  const { type, payload } = action;
  switch (type) {
    case SET_USER_ACCOUNT_INFO:
      return {
        ...state,
        first_name: payload.first_name,
        last_name: payload.last_name,
        email: payload.email,
        phone_number: payload.phone_number,
        is_admin: payload.is_admin
      };
    default:
      return state;
  }
}

