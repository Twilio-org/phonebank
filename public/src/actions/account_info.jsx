import axios from 'axios';
import { SET_USER_ACCOUNT_INFO } from '../reducers/account_info';

export function setAccountInfo(user) {
  return {
    type: SET_USER_ACCOUNT_INFO,
    payload: user
  };
}

export function fetchUser(id) {
  return dispatch => axios.get(`/users/${id}`, {
    headers: { 'Authorization': ` JWT ${localStorage.getItem('auth_token')}` }
  })
  .then((res) => {
    const userData = res.data;
    console.log(userData, 'on fetch user')
    return dispatch(setAccountInfo(userData));
  })
  .catch((err) => {
    const customError = {
      message: `error fetching user info in account_info action fetchUserInfo: ${err}`,
      name: 'user info post request from account_info component'
    };
    throw customError;
  });
}
