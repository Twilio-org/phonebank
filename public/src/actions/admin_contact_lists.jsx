import axios from 'axios';
import { SET_CAMPAIGN_FORM_CONTACT_LIST, SET_CURRENT_CONTACT_LIST } from '../reducers/admin_contact_lists';

export function setContactListOptions(contactLists) {
  return {
    type: SET_CAMPAIGN_FORM_CONTACT_LIST,
    payload: contactLists
  };
}

export function setCurrentContactList(contactListObj) {
  return {
    type: SET_CURRENT_CONTACT_LIST,
    payload: contactListObj
  };
}


export function fetchAllContactLists() {
  return dispatch => axios.get('/contactLists')
    .then((res) => {
      const contactLists = res.data;
      dispatch(setContactListOptions(contactLists));
      return contactLists;
    })
    .catch((err) => {
      const customError = {
        message: `error fetching contact lists: ${err}`,
        name: 'contact lists get request from campaignPage component'
      };
      throw customError;
    });
}
