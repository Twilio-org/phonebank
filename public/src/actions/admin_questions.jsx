import axios from 'axios';
import { destroy } from 'redux-form';
import { SET_QUESTIONS, SET_QUESTION_CURRENT } from '../reducers/admin_questions';

export function setQuestionList(questionList) {
  return {
    type: SET_QUESTIONS,
    payload: questionList
  };
}

export function setCurrentQuestion(questionDataObj) {
  return {
    type: SET_QUESTION_CURRENT,
    payload: questionDataObj
  };
}

export function fetchAllQuestions() {
  return dispatch => axios.get('/questions', {
    headers: { Authorization: ` JWT ${localStorage.getItem('auth_token')}` }
  })
  .then((questions) => {
    const { data: questionsList } = questions;
    return dispatch(setQuestionList(questionsList.sort((a, b) => (a.id - b.id))));
  })
  .catch((err) => {
    const customError = {
      message: `error fetching questions: ${err}`,
      name: 'question get request from script_form component'
    };
    throw customError;
  });
}

export function fetchQuestion(id) {
  return dispatch => axios.get(`/questions/${id}`, {
    headers: { Authorization: ` JWT ${localStorage.getItem('auth_token')}` }
  })
  .then((res) => {
    const { data: questionData } = res;
    // TO-DO Find a better way to handle this data
    questionData.responses = questionData.responses ? questionData.responses.split(',') : '';
    return dispatch(setCurrentQuestion(questionData));
  })
  .catch((err) => {
    const customError = {
      message: `error fetching question info in questions action fetchQuestion: ${err}`,
      name: 'question info get request'
    };
    throw customError;
  });
}

export function createQuestion(questionInfo, history) {
  let { responses } = questionInfo;
  const { title, description, type } = questionInfo;
  // TO-DO: Find a better way to handle question response data
  responses = responses ? Object.keys(responses).map(key => responses[key]).join(',') : '';

  return dispatch => axios.post('/questions', {
    title,
    description,
    type,
    responses
  })
  .then((res) => {
    history.goBack();
    dispatch(destroy('CreateQuestion'));
    return res;
  })
  .catch((err) => {
    const customError = {
      message: `error in creating new questions to db: ${err}`,
      name: 'createQuestion function error in actions/questions.jsx'
    };
    throw customError;
  });
}
