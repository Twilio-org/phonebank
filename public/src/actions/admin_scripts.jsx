import axios from 'axios';
import { SET_SCRIPTS, SET_SCRIPT_CURRENT } from '../reducers/admin_scripts';

export function setScriptsList(scriptList) {
  return {
    type: SET_SCRIPTS,
    payload: scriptList
  };
}

export function setCurrentScript(scriptDataObj) {
  return {
    type: SET_SCRIPT_CURRENT,
    payload: scriptDataObj
  };
}

export function fetchAllScripts() {
  return dispatch => axios.get('/scripts', {
    headers: { Authorization: ` JWT ${localStorage.getItem('auth_token')}` }
  })
  .then((scripts) => {
    console.log(scripts, '##### succesfully grabed scrpts')
    const { data: scriptList } = scripts;
    return dispatch(setScriptsList(scriptList));
  })
  .catch((err) => {
    console.log('error fetching all scripts: ', err);
  });
}
