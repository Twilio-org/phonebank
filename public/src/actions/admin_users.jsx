import axios from 'axios';
import { SET_USERS, SET_USER_CURRENT } from '../reducers/admin_users';

export function setUserList(usersList) {
  return {
    type: SET_USERS,
    payload: usersList
  };
}

export function setCurrentUser(userObject) {
  return {
    type: SET_USER_CURRENT,
    payload: userObject
  };
}

export function fetchAllUsers(currentUserId) {
  return dispatch => axios.get('/users', {
    headers: { Authorization: ` JWT ${localStorage.getItem('auth_token')}` }
  })
  .then((users) => {
    const { data: usersList } = users;
    const modifiedUsersList = usersList.map((userObject) => {
      const { is_admin, is_banned, is_active } = userObject;
      return {
        ...userObject,
        is_admin: JSON.stringify(is_admin),
        is_banned: JSON.stringify(is_banned),
        is_active: JSON.stringify(is_active)
      };
    });
    // remove the current user from the user's list... so they don't accidentally
    // change their status from admin to volunteer and have all the bugs happen
    const filteredUsersList = modifiedUsersList.filter((object) => {
      const { id } = object;
      return id !== currentUserId;
    });
    return dispatch(setUserList(filteredUsersList));
  })
  .catch(err => console.log('problem fetching all users from db in action "fetchAllUsers"', err));
}

export function adminUpdateUserInfo(id, target, newValue, currentUserId, fetchAll = fetchAllUsers) {
  const params = {};
  params[target] = JSON.stringify(newValue);
  return dispatch => axios.put(`/users/${id}/manage`, params,
    {
      headers: { Authorization: ` JWT ${localStorage.getItem('auth_token')}` }
    }
  )
  .then(() => {
    dispatch(fetchAll(currentUserId));
  })
  .catch(err => console.log('error with user management update action: ', err));
}
