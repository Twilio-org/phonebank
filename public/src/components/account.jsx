import React, { Component } from 'react';
// import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import { Button } from 'react-bootstrap';

import { fetchUser, deleteUser } from '../actions/account_info';

class AccountPage extends Component {
  constructor(props) {
    super(props);
    this.onDeleteClick = this.onDeleteClick.bind(this);
  }
  componentDidMount() {
    const { id } = this.props.auth;
    this.props.fetchUser(id);
  }

  onDeleteClick() {
    const { id } = this.props.auth;
    console.log(id, 'user id on delete click')
    const { history } = this.props;
    console.log(this.props, 'meow');
    this.props.deleteUser(id, history);
  }

  render() {
    const { account_info } = this.props;
    const { id } = this.props.auth;
    const { first_name, last_name, email, phone_number } = account_info;

    return (
      <div>
        <h6>Name:</h6>
        <h3>{!!account_info ? `${first_name} ${last_name}` : ''}</h3>
        <h6>Email:</h6>
        <h3>{!!account_info ? email : ''}</h3>
        <h6>Phone:</h6>
        <h3>{!!account_info ? phone_number : ''}</h3>
        <div className="text-xs-right">
          <Link className="btn btn-primary" to={`/account/${id}/edit`}>
            Edit Account
          </Link>
        </div>
        <div>
          <Button bsStyle="primary" onClick={this.onDeleteClick}>Delete Account</Button>
        </div>
      </div>
    );
  }
}

// AccountPage.propTypes = {
//   fetchUser: PropTypes.func,
//   auth: PropTypes.obj,
//   account_info: PropTypes.obj
// };

// AccountPage.defaultProps = {
//   fetchUser: () => 'meow',
//   auth: { id: 1 },
//   account_info: {
//     first_name: 'meow',
//     last_name: 'meows',
//     email: 'meow@meow.com',
//     phone_number: '12342345'
//   }
// };

function mapStateToProps(state) {
  return { auth: state.auth, account_info: state.account_info };
}

export default withRouter(
  connect(mapStateToProps, { fetchUser, deleteUser })(AccountPage)
);
