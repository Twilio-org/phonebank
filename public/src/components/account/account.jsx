import React, { Component } from 'react';
// import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import { Button, ButtonToolbar, PageHeader } from 'react-bootstrap';

import { fetchUser, deleteUser } from '../../actions/users';

export class AccountPage extends Component {
  constructor(props) {
    super(props);
    this.onDeleteClick = this.onDeleteClick.bind(this);
    // DO NOT DELETE THE FOLLOWING LINES!!!!!
    // ['onDeleteClick'].forEach((func) => {
    //   this[func] = this[func].bind(this);
    // });
  }

  componentDidMount() {
    const { id } = this.props.auth;
    this.props.fetchUser(id);
  }

  onDeleteClick() {
    const { id } = this.props.auth;
    const { history } = this.props;
    this.props.deleteUser(id, history);
  }

  render() {
    const { account_info } = this.props;
    const { id } = this.props.auth;
    const { first_name, last_name, email, phone_number, is_admin } = account_info;

    return (
      <div className="admin-form">
        <PageHeader>Account</PageHeader>
        <h6>Name:</h6>
        <h3>{account_info ? `${first_name} ${last_name}` : ''}</h3>
        <h6>Email:</h6>
        <h3>{account_info ? email : ''}</h3>
        <h6>Phone:</h6>
        <h3>{account_info ? phone_number : ''}</h3>
        <ButtonToolbar>
          <Link className="btn btn-primary" to={is_admin ? `/admin/account/${id}/edit` : `/volunteers/account/${id}/edit`}>
            Edit Account
          </Link>
          <Button bsStyle="danger" onClick={this.onDeleteClick}>Delete Account</Button>
        </ButtonToolbar>
      </div>
    );
  }
}

function mapStateToProps(state) {
  return { auth: state.auth, account_info: state.account_info };
}

export const ConnectedAccountPage = withRouter(
  connect(mapStateToProps, { fetchUser, deleteUser })(AccountPage)
);
