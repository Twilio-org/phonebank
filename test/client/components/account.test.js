import React from 'react';
import { mount, shallow } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { AccountPage } from '../../../public/src/components/account/account';
import { fetchUser } from '../../../public/src/actions/account_info';
import deleteUser from '../../../public/src/actions/edit_account';

jest.mock('../../../public/src/actions/edit_account', () => jest.fn());

const mockStore = {
  auth: {
    id: 31
  },
  accountInfo: {
    first_name: 'Harry',
    last_name: 'Potter',
    email: 'goldensnitch@hogwarts.com',
    phone_number: '12224448888'
  }
};

describe('<AccountPage />', () => {
  const { accountInfo, auth } = mockStore;
  const wrapper = mount(
    <MemoryRouter>
      <AccountPage
        account_info={accountInfo}
        auth={auth}
        fetchUser={fetchUser}
        deleteUser={deleteUser}
      />
    </MemoryRouter>
  );
  it('renders correctly', () => {
    const tree = renderer.create(
      <MemoryRouter>
        <AccountPage
          account_info={accountInfo}
          auth={auth}
          fetchUser={fetchUser}
          deleteUser={deleteUser}
        />
      </MemoryRouter>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should render 3 div elements', () => {
    expect(wrapper.find('div').length).toBe(3);
  });
  it('should render a Link component', () => {
    expect(wrapper.find('Link').length).toBe(1);
  });
});

// Props + methods:
describe('AccountPage props and methods:', () => {
  const props = {
    auth: {
      id: 31
    },
    account_info: {
      first_name: 'Harry',
      last_name: 'Potter',
      email: 'goldensnitch@hogwarts.com',
      phone_number: '12224448888'
    },
    deleteUser,
    fetchUser,
    history: 'history'
  };
  const wrapper = shallow(<AccountPage {...props} />).instance();
  const { onDeleteClick } = wrapper;
  describe('click handler for user Delete (onDeleteClick)', () => {
    it('should have a method named onDeleteClick', () => {
      expect(typeof onDeleteClick).toBe('function');
    });
    it('should call this.props.deleteUser only once on button click', () => {
      wrapper.onDeleteClick();
      const numberOfMockCalls = deleteUser.mock.calls.length;
      expect(deleteUser).toHaveBeenCalledTimes(1);
      expect(numberOfMockCalls).toBe(1);
    });
    it('should call this.props.deleteUser with two arguments: id (31) and history', () => {
      const mockCallsArray = deleteUser.mock.calls[0];
      expect(mockCallsArray.length).toBe(2);
      expect(mockCallsArray[0]).toBe(31);
      expect(mockCallsArray[1]).toBe('history');
    });
  });

  describe('has props from the store connection', () => {
    describe('props should contain auth', () => {
      it('should receive auth which is an object', () => {
        expect(typeof props.auth).toBe('object');
      });
      it('auth should have a property \'id\' with a value of 31', () => {
        expect(props.auth.id).toBeDefined();
        expect(props.auth.id).toBe(31);
      });
    });
    describe('props should contain account_info', () => {
      it('should recieve account_info which is an object', () => {
        expect(typeof props.account_info).toBe('object');
      });
      it('account info should have 4 properties: first_name, last_name, email, phone_number', () => {
        const accountInfoKeys = Object.keys(props.account_info);
        expect(accountInfoKeys.length).toBe(4);
        accountInfoKeys.forEach((key) => {
          expect(Object.prototype.hasOwnProperty.call(props.account_info, key)).toBe(true);
        });
      });
    });
  });
});
