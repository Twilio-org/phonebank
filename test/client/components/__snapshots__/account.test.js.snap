// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccountPage /> renders correctly 1`] = `
<div
  className="admin-form"
>
  <div
    className="page-header"
  >
    <h1>
      Account
    </h1>
  </div>
  <h6>
    Name:
  </h6>
  <h3>
    Harry Potter
  </h3>
  <h6>
    Email:
  </h6>
  <h3>
    goldensnitch@hogwarts.com
  </h3>
  <h6>
    Phone:
  </h6>
  <h3>
    12224448888
  </h3>
  <div
    className="btn-toolbar"
    role="toolbar"
  >
    <a
      className="btn btn-primary"
      href="/volunteers/account/31/edit"
      onClick={[Function]}
    >
      Edit Account
    </a>
    <button
      className="btn btn-danger"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Delete Account
    </button>
  </div>
</div>
`;
